### PR TITLE
Add new parameters to taskdefinition module - network_mode and task_r…

### PR DESCRIPTION
**Pull request by wimnat**
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

ecs_taskdefinition
##### ANSIBLE VERSION

2.2
##### SUMMARY

Resolves #3015 

Adds two new parameters to the module:

task_role_arn 
network_mode

These are new features in AWS

### CODE REVIEW

**Comment by wimnat** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/3016

**Review by Java1Guy — APPROVED** [review]

**Review by Java1Guy — APPROVED** [review]

**Comment by gundalow** [comment]
You appear to have made this file executable, please revert that

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 1]
```diff
@@ -51,6 +51,18 @@
             - A list of containers definitions
         required: False
         type: list of dicts with container definitions
+    network_mode:
+        description:
+            - The Docker networking mode to use for the containers in the task.
+        required: false
+        default: bridge
+        choices: [ 'bridge', 'host', 'none' ]
+        version_added: 2.2
```
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]: This needs to be a string: `"2.3"`

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py [hunk 2]
```diff
@@ -51,6 +51,18 @@
...
+    task_role_arn:
+        description:
+            - The Amazon Resource Name (ARN) of the IAM role that containers in this task can assume. All containers in this task are granted the permissions that are specified in this role.
+        required: false
+        version_added: 2.2
```
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]: This needs to be a string: `"2.3"`

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

**Comment by wimnat** [comment]
ready_for_review

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Given maintainer @Java1Guy has approved this I'll merge

**Comment by kkarimi** [comment]
@wimnat I think there is a bug here but not sure, when I try with this definition:
 ```
 - name: Registrator - Create Task Definition
  ecs_taskdefinition:
    state: present
    family: "registrator"
    network_mode: host
    containers:
      - name: registrator
        image: "gliderlabs/registrator:master"
        memory: 10
        portMappings:
          - containerPort: 9998
            hostPort: 9998
            protocol: tcp
        mountPoints:
          - sourceVolume: docker-sock
            containerPath: /var/run/docker.sock
            readOnly: true
 ```

I get:
 ```
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_name": "ecs_taskdefinition"
    }, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/9c/3f5ffpbx6p977jwgv3lj3d8c0000gn/T/ansible_sq3saL/ansible_module_ecs_taskdefinition.py\", line 383, in <module>\n    main()\n  File \"/var/folders/9c/3f5ffpbx6p977jwgv3lj3d8c0000gn/T/ansible_sq3saL/ansible_module_ecs_taskdefinition.py\", line 352, in main\n    volumes)\n  File \"/var/folders/9c/3f5ffpbx6p977jwgv3lj3d8c0000gn/T/ansible_sq3saL/ansible_module_ecs_taskdefinition.py\", line 176, in register_task\n    self.module.fail_json(msg=e.message, **camel_dict_to_snake_dict(e.response))\nNameError: global name 'camel_dict_to_snake_dict' is not defined\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
```

which seems to be related to [this line](https://github.com/ansible/ansible/pull/19493/files#diff-30a8e3afd34512ab4a4fed7f5ecfa9a4R176).

**Comment by wimnat** [comment]
Hmm what version of ansible are you using? That function has been in since
2.1 I think

On 20 Dec. 2016 12:27, "Kash" <notifications@github.com> wrote:

@wimnat <https://github.com/wimnat> I think there is a bug here but not
sure, when I try with this definition:

- name: Registrator - Create Task Definition
 ecs_taskdefinition:
   state: present
   family: "registrator"
   network_mode: host
   containers:
     - name: registrator
       image: "gliderlabs/registrator:master"
       memory: 10
       portMappings:
         - containerPort: 9998
           hostPort: 9998
           protocol: tcp
       mountPoints:
         - sourceVolume: docker-sock
           containerPath: /var/run/docker.sock
           readOnly: true

I get:

fatal: [localhost]: FAILED! => {
   "changed": false,
   "failed": true,
   "invocation": {
       "module_name": "ecs_taskdefinition"
   },
   "module_stderr": "Traceback (most recent call last):\n  File
\"/var/folders/9c/3f5ffpbx6p977jwgv3lj3d8c0000gn/T/ansible_sq3saL/ansible_module_ecs_taskdefinition.py\",
line 383, in <module>\n    main()\n  File
\"/var/folders/9c/3f5ffpbx6p977jwgv3lj3d8c0000gn/T/ansible_sq3saL/ansible_module_ecs_taskdefinition.py\",
line 352, in main\n    volumes)\n  File
\"/var/folders/9c/3f5ffpbx6p977jwgv3lj3d8c0000gn/T/ansible_sq3saL/ansible_module_ecs_taskdefinition.py\",
line 176, in register_task\n    self.module.fail_json(msg=e.message,
**camel_dict_to_snake_dict(e.response))\nNameError: global name
'camel_dict_to_snake_dict' is not defined\n",
   "module_stdout": "",
   "msg": "MODULE FAILURE",
   "rc": 0
}

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/ansible/ansible/pull/19493#issuecomment-268231076>, or mute
the thread

**Comment by wimnat** [comment]
Actually I see the problem. The import statement doesn't include the function. I'll submit a PR tonight. Not sure how this passed my testing :/

**Comment by kkarimi** [comment]
@wimnat I am running against develop branch, so that must be it, thanks 👍 

- edit: I made a new PR #19560 to fix it, please feel free to review

**Pull request merged**